thrift/src/main/java/com/linecorp/armeria/common/thrift/text/StructContext.java

```diff
@@ -132,35 +138,85 @@ protected TField getTFieldByName(String name) throws TException {
-
-        for (StackTraceElement f : frames) {
-            final String className = f.getClassName();
-
+        if (SystemInfo.javaVersion() >= 9) {
+            Optional<Class> result = Optional.empty();
             try {
-                final Class<?> clazz = Class.forName(className);
-
-                // Note, we need to check
-                // if the class is abstract, because abstract class does not have metaDataMap
-                // if the class has no-arg constructor, because FieldMetaData.getStructMetaDataMap
-                //   calls clazz.newInstance
-                if (isTBase(clazz) && !isAbstract(clazz) && hasNoArgConstructor(clazz)) {
-                    return clazz;
-                }
+                final Class<?> stackWalker = Class.forName("java.lang.StackWalker");
+                final Object stackWalkerInstance = stackWalker
+                        .getDeclaredMethod("getInstance")
+                        .invoke(null);
+                final Method walk = stackWalkerInstance.getClass().getDeclaredMethod("walk", Function.class);
+                result = (Optional<Class>) walk.invoke(
+                        stackWalkerInstance,
+                        (Function<Stream<Object>, Optional<Class>> ) s -> streamProcessor(s));
+            } catch (Throwable t) {
+                // @TODO: msg
+                log.debug("walwlawla", "StackWalker.getInstance.walk", t);
+            }
 
-                if (isTApplicationException(clazz)) {
-                    return clazz;
-                }
+            if (result.isPresent()) {
+                return result.get();
+            }
+        } else { // below java version 8
+            final StackTraceElement[] frames =
+                    Thread.currentThread().getStackTrace();
+            final Optional<Class> result = streamProcessor(Stream.of(frames));
 
-                if (isTApplicationExceptions(clazz)) {
-                    return TApplicationException.class;
-                }
-            } catch (ClassNotFoundException ex) {
-                log.warn("Can't find class: " + className, ex);
+            if (result.isPresent()) {
+                return result.get();
             }
         }
-        throw new RuntimeException("Must call (indirectly) from a TBase/TApplicationException object.");
+        throw  new RuntimeException("Must call (indirectly) from a TBase/TApplicationException object.");
+    }
+
+    private static Optional<Class> streamProcessor(Stream<Object> stream) {
+        return stream.map(StructContext::getClassName)
+                .map(StructContext::getClassByName)
+                .filter(Optional::isPresent)
+                .map(StructContext::matchClass)
+                .filter(Objects::nonNull)
+                .findFirst();
+    }
+
+    private static String getClassName(Object stackFrame) {
+        try {
+            final Method getClassName = stackFrame.getClass().getDeclaredMethod("getClassName");
```

Comment:
This is inefficient. Use `StackTraceElement.getClassName()` directly for the default implementation and `MethodHandle` for the Java 9 implementation. Once you introduce some class hierarchy as I advised, you could provide two different `getClassName()` implementations.